README.pod
=pod

=head1 NAME

README for dta-tokwrap - programs, scripts, and perl modules for DTA XML corpus tokenization

=cut

##======================================================================
=pod

=head1 DESCRIPTION

This package contains various utilities for
tokenization of DTA "base-format" XML documents.
see L</INSTALLATION> for requirements and installation instructions,
see L</USAGE> for a brief introduction to the high-level command-line interface,
and
see L</TOOLS> for an overview of the individual tools included in this distribution.

=cut


##======================================================================
=pod

=head1 INSTALLATION

=cut

##--------------------------------------------------------------
=pod

=head2 Requirements

=cut

##~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
=pod

=head3 C Libraries

=over 4

=item expat

tested version(s): 1.95.8, 2.0.1

=item libxml2

tested version(s): 2.7.3, 2.7.8

=item libxslt

tested version(s): 1.1.24, 1.1.26

=back

=cut

##~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
=pod

=head3 Perl Modules

See F<DTA-TokWrap/README.txt> for a full list of required
perl modules.

=cut

##~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
=pod

=head3 Development Tools

=over 4

=item C compiler

tested version(s): gcc / linux: v4.3.3, 4.4.6

=item GNU flex (development only)

tested version(s): 2.5.33, 2.5.35

Only needed if you plan on making changes to the lexer sources.

=item GNU autoconf (SVN only)

tested version(s): 2.61, 2.67

Required for building from SVN sources.

=item GNU automake (SVN only)

tested version(s): 1.9.6, 1.11.1

Required for building from SVN sources.

=back

=cut

##--------------------------------------------------------------
=pod

=head2 Building from SVN

To build this package from SVN sources, you must first run the shell command:

 bash$ sh ./autoreconf.sh

from the distribution root directory B<BEFORE> running F<./configure>.
Building from SVN sources requires additional development tools to present
on the build system.  Then, follow the instructions in L</"Building from Source">.

=cut

##--------------------------------------------------------------
=pod

=head2 Building from Source

To build and install the entire package, issue the following commands to the shell:

 bash$ cd dta-tokwrap-0.01   # (or wherever you unpacked this distribution)
 bash$ sh ./configure        # configure the package
 bash$ make                  # build the package
 bash$ make install          # install the package on your system

More details on the top-level installation process can be found in
the file F<INSTALL> in the distribution root directory.

More details on building and installing the DTA::TokWrap perl module included in this distribution
can be found in the F<perlmodinstall(1)> manpage.

=cut

##======================================================================
=pod

=head1 USAGE

The perl program L<dta-tokwrap.perl|/dta-tokwrap.perl> installed from the F<DTA-TokWrap/>
distribution subdirectory provides a flexible high-level command-line interface
to the tokenization of DTA XML documents.

=cut

##--------------------------------------------------------------
=pod

=head2 Input Format

The L<dta-tokwrap.perl|dta-tokwrap.perl> script takes as its input DTA "base-format" XML files,
which are simply (TEI-conformant) UTF-8 encoded XML files with one C<E<lt>cE<gt>>
element per character:

=over 4

=item *

the document B<MUST> be encoded in UTF-8,

=item *

all text nodes to be tokenized should be descendants of a C<E<lt>textE<gt>> element,
and may optionally be immediate daughters of a C<E<lt>cE<gt>> element
(XPath C<//text//text()|//text//c/text()>). C<E<lt>cE<gt>> elements may not be nested.

Prior to dta-tokwrap v0.38, C<E<lt>cE<gt>> elements were required.

=back

=cut

##--------------------------------------------------------------
=pod

=head2 Example: Tokenizing a single XML file

Assume we wish to tokenize a single DTA "base-format" XML file F<doc1.xml>.
Issue the following command to the shell:

 bash$ dta-tokwrap.perl doc1.xml

... This will create the following output files:

=over 4

=item F<doc1.t.xml>

"Master" tokenizer output file encoding sentence boundaries, token boundaries,
and tokenizer-provided token analyses.  Source for various stand-off annotation formats.
This format can also be passed directly to and from the L<DTA::CAB(3pm)|DTA::CAB>
analysis suite using the L<DTA::CAB::Format::XmlNative(3pm)|DTA::CAB::Format::XmlNative>
formatter class.

=back

=cut

##--------------------------------------------------------------
=pod

=head2 Example: Tokenizing multiple XML files

Assume we wish to tokenize a corpus of three DTA "base-format" XML files
F<doc1.xml>, F<doc2.xml>, and F<doc3.xml>.
This is as easy as:

 bash$ dta-tokwrap.perl doc1.xml doc2.xml doc3.xml

For each input document specified on the command line,
master output files and stand-off annotation files will be created.

See L<"the dta-tokwrap.perl manpage"|dta-tokwrap.perl> for more details.

=head2 Example: Tracing execution progess

Assume we wish to tokenize a large corpus of XML input files F<doc*.xml>,
and would like to have some feedback on the progress of the
tokenization process.
Try:

 bash$ dta-tokwrap.perl -verbose=1 doc*.xml

or:

 bash$ dta-tokwrap.perl -verbose=2 doc*.xml

or even:

 bash$ dta-tokwrap.perl -traceAll doc*.xml

=cut

##======================================================================
=pod

=head1 TOOLS

This section provides a brief overview of the individual tools included
in the dta-tokwrap distribution.

=cut

##--------------------------------------------------------------
=pod

=head2 Perl Scripts & Programs

The perl scripts and programs included with this distribution are installed
by default in F</usr/local/bin> and/or wherever your perl installs
scripts by default (e.g. in C<`perl -MConfig -e 'print $Config{installsitescript}'`>).

=over 4

=item dta-tokwrap.perl

Top-level wrapper script for document tokenization
using the L<DTA::TokWrap|DTA::TokWrap> perl API.

=item dtatw-add-c.perl

Script to insert C<E<lt>cE<gt>> elements and/or
C<xml:id> attributes for such elements into an
XML document which does not yet contain them.
Guaranteed not to clobber any existing //c IDs.

//c/@xml:id attributes are generated by a simple document-global counter
("c1", "c2", ..., "c65536").

See L<"the dtatw-add-c.perl manpage"|dtatw-add-c.perl> for more details.

=item dtatw-cids2local.perl

Script to convert C<//c/@xml:id> attributes to page-local encoding.
Never really used.

See L<"the dtatw-cids2local.perl manpage"|dtatw-cids2local.perl> for more details.

=item dtatw-add-ws.perl

Script to splice C<E<lt>sE<gt>> and C<E<lt>wE<gt>> elements encoded from a standoff (.t.xml or .u.xml) XML file
into the I<original> "base-format" (.chr.xml) file, producing a .cws.xml file.
A tad too generous with partial word segments, due to strict adjacency and boundary criteria.

In earlier versions of dta-tokwrap, this functionality was split between the scripts
C<dtatw-add-w.perl> and C<dtatw-add-s.perl>, which required only an I<id-compatible>
base-format (.chr.xml) file as the splice target.  As of dta-tokwrap v0.35, the splice target
base-format file must be I<original> source file itself, since the current implementation
uses byte offsets to perform the splice.

See L<"the dtatw-add-ws.perl manpage"|dtatw-add-ws.perl> for more details.

=item dtatw-splice.perl

Script to splice generic standoff attributes and/or content into a base file;
useful e.g. for merging flat DTA::CAB standoff analyses into TEI-structured
*.cws.xml files.

See L<"the dtatw-splice.perl manpage"|dtatw-splice.perl> for more details.

=item dtatw-get-ddc-attrs.perl

Script to insert DDC-relevant attributes extracted from a base file into a *.t.xml file,
producing a pre-DDC XML format file (by convention *.ddc.t.xml, a subset of the *.t.xml
format).

See L<"the dtatw-get-ddc-attrs.perl manpage"|dtatw-get-ddc-attrs.perl> for more details.

=item dtatw-get-header.perl

Simple script to extract a single header element from an XML file (e.g. for later
inclusion in a DDC XML format file).

See L<"the dtatw-get-header.perl manpage"|dtatw-get-header.perl> for more details.

See L<"the dtatw-get-header.perl manpage"|dtatw-get-header.perl> for more details.

=item dtatw-pn2p.perl

Script to conver insert E<lt>pE<gt>...E<lt>/pE<gt> wrappers for C<//s/@pn> key attributes
in "flat" *.t.xml files.

=item dtatw-xml2ddc.perl

Script to convert *.ddc.t.xml files and optional headers to DDC-XML format.

See L<"the dtatw-xml2ddc.perl manpage"|dtatw-xml2ddc.perl> for more details.

=item dtatw-t-check.perl

Simple script to check consistency of tokenizer output (*.t) offset + length
fields with input (*.txt) file.

=item dtatw-add-c.perl

Script to add C<E<lt>cE<gt>> elements to an XML document which does
not already contain them.
Not really useful as of dta-tokwrap v0.38.

=item dtatw-rm-c.perl

Script to remove C<E<lt>cE<gt>> elements from an XML document.
Regex hack, fast but not exceedingly robust, use with caution.
See also L</"dtatw-rm-c.xsl">

=item dtatw-rm-w.perl

Fast regex hack to remove C<E<lt>wE<gt>> elements from an XML document

=item dtatw-rm-s.perl

Fast regex hack to remove C<E<lt>sE<gt>> elements from an XML document.

=item dtatw-rm-lb.perl

Script to remove C<E<lt>lbE<gt>> (line-break) elements from an XML document,
replacing them with newlines.
Regex hack, fast but not robust, use with caution.
See also L</"dtatw-rm-lb.xsl">

=item dtatw-lb-encode.perl

Encodes newlines under //text//text() in an XML document as C<E<lt>lbE<gt>> (line-break) elements
using high-level file heuristics only.
Regex hack, fast but not robust, use with caution.
See also L</"dtatw-ensure-lb.perl">, L</"dtatw-add-lb.xsl">, L</"dtatw-rm-lb.perl">.

=item dtatw-ensure-lb.perl

Script to ensure that all //text//text() newlines in an XML document are explicitly encoded
with C<E<lt>lbE<gt>> (line-break) elements, using optional file-, element-,
and line-level heuristics.
Robust but slow, since it actually parses XML input documents.
See also L</"dtatw-lb-encode.perl">, L</"dtatw-add-lb.xsl">, L</"dtatw-rm-lb.perl">.

=item dtatw-tt-dictapply.perl

Script to apply a type-"dictionary" in one-word-per-line (.tt) format to a
token corpus in one-word-per-line (.tt) format.  Especially useful together with
standard UNIX utilities such as cut, grep, sort, and uniq.

=item dtatw-cabtt2xml.perl

Script to convert DTA::CAB::Format::TT (one-word-per-line with variable analysis
fields identified by conventional prefixes) files to expanded .t.xml format used
by dta-tokwrap.  The expanded format should be identical to that used by the
DTA::CAB::Format::Xml class.  See also L<dtatw-txml2tt.xsl>.

=item file-substr.perl

Script to extract a portion of a file,
specified by byte offset and length.
Useful for debugging index files created by other tools.

=back

=cut

##--------------------------------------------------------------
=pod

=head2 GNU make build system template

The distribution directory F<make/> contains a "template"
for using GNU F<make> to
organizing the conversion of large corpora with
the dta-tokwrap utilities.  This is useful because:

=over 4

=item *

F<make>'s intuitive, easy-to-read syntax provides a

ChangeLog
##-*- Mode: ChangeLog; coding: utf-8; -*-

v0.91 Mon, 22 Jul 2019 13:57:06 +0200 moocow
	* added dtatw-xml-depth : get maximum element nesting depth for XML file(s)

v0.90 Mon, 11 Mar 2019 12:36:38 +0100 moocow
	* mkbx0: ignore 'metamark'

v0.89 Fri, 22 Feb 2019 09:31:13 +0100 moocow
	* top-level Makefile.PL tweaks
	  - META.* re-generation fixes
	  - declare lack of support for win32

v0.88 Thu, 21 Feb 2019 21:18:49 +0100 moocow
	* re-added missing top-level Makefile.PL (seems to have gotten lost in svn merge)
	* set $ENV{PERL}=$^X in Makefile.PL before calling ./configure
	  - should fix bogus failures from non-default perls (e.g. cpantesters ae09febc-353f-11e9-a0cc-de79a423f08d)
	  - many thanks to Slaven Rezić for spotting the problem

v0.87 Thu, 21 Feb 2019 15:21:05 +0100 moocow
	* first non-dev cpan release

v0.86 Wed, 20 Feb 2019 08:31:17 +0100 moocow
	* refactored dta-tokwrap distribution for cpanm- & cpantesters-friendliness

v0.85 Tue, 06 Nov 2018 15:54:02 +0100 moocow
	* scripts/dtatw-sanitize-header.perl: added length-based trimming for sanitized bibl fields (default: -max-bibl-length=256)
	* scripts/dtatw-get-ddc-attrs.perl: removed 'left' context-element for 'xc' attribute (mantis #31734)

v0.84 Thu, 13 Sep 2018 14:39:19 +0200 moocow
	* added dtatw-fast-ddc-attrs.perl: fast minimal attribute extraction (//w/@ws only)

v0.83 Wed, 05 Sep 2018 11:05:56 +0200 moocow
	* added dtatw-sanitize-header.perl support for user-specified XPaths

v0.82 Fri, 10 Aug 2018 10:07:11 +0200 moocow
	* added TCF->TEI decoding support for TEI att.linguistic attributes //w/(@lemma|@pos|@norm|@join)
	  - uses new processor module 'txmlanno': in-place update of *.t.xml
	  - optional: only used if tcfdecode option 'att.linguistic' is set
	  - wrapped by new tei-tcf web-service v0.06 form parameter 'lingattrs'

v0.81 Fri, 13 Apr 2018 10:53:17 +0200 moocow
	* removed diagnostic comments for non-initial chained material in Processor::mkbx0::chain_stylestr()
	  - fixes mantis bug #26675: comments caused XSL transform to choke in Processor::mkbx0 for xml:ids containing trailing hyphens

v0.80 Tue, 03 Apr 2018 13:57:33 +0200 moocow
	* allow TCF->TEI decoding even without a TCF 'tokens' layer (expensive no-op)
	* added TCF<->TEI encoding/decoding example in top-level README

v0.79 Wed, 08 Nov 2017 12:20:54 +0100 moocow
	* dtatw-sanitize-header.perl for 'rsc' corpus tweaks (//idno fallback XPaths)

v0.78 Wed, 26 Jul 2017 12:57:07 +0200 moocow
	* dtatw-sanitize-header.perl for 'rem' corpus tweaks (date string sanitation heuristics)

v0.77 Tue, 21 Mar 2017 14:40:32 +0100 moocow
	* added dtatw-percent-(encode|decode).perl : "%" <-> "$%$" escaping for use with waste tokenizer >= v2.0.15-1

v0.76 Wed, 25 Jan 2017 11:05:49 +0100 moocow
	* changed dtatw-get-ddc-attrs.perl @rendition parsing (ALL -> ANY); related to mantis bug #18392

v0.75 2016-11-09  moocow
	* fixed handling of -po=waste=PATH for 'auto' tokenizer class

v0.74 2016-11-01  moocow
	* updated default tcf textSource type (again)

v0.73 2016-09-23  moocow
	* added character-offset mode to file-substr.perl (expensive, buffers whole file)

v0.73 2016-06-07  moocow
	* added tei2spliced target
	* added tei2spliced target
	* updated docs
	* added dta-tokwrap.perl -waste-dir option

v0.72 2016-05-12  moocow
	* better docs for dtatw-sanitize-header
	* improved basename guessing in dtatw-sanitize-header.perl for header-less dta files

v0.71 2015-11-12  moocow
	* dtatw-lb-encode.perl fixes: \R regex was splitting UTF8 characters --> malformed xml

v0.71 2015-08-19  moocow
	* added fast regex hack dtatw-lb-encode.perl (for dstar build)
	* added dtatw-ensure-lb.perl : insert <lb/> where tokwrap expects it

v0.69 2015-07-23  moocow
	* dtatw-sanitize-header.perl: auto-normalize whitespace in fields
	 - fixes broken DDC return values involving TABs in metadata

v0.68 2015-06-16  moocow
	* added aux-db support to dtatw-sanitize-header.perl

v0.67 2015-06-15  moocow
	* dtatw-sanitize-header compat fixes
	* dtatw-sanitize-header.perl: new canonical XPaths for dtaid, dtadir

v0.66 2015-06-10  moocow
	* added dtatw-insert-header.perl : header splicing (e.g. for metadata tweaking during dstar con-import)

v0.65 2015-03-11  moocow
	* re-serialize <metamark> a la <note> and friends

v0.64 2015-03-06  moocow
	* mkbx0: no whitespace before sb, wb elements (for dta ws attribute)

v0.63 2015-02-18  moocow
	* added div_TYPE components to 'xc' (ddc 'con' field)

v0.63 2015-02-09  moocow
	* ignore //del in mkbx0 (fixes mantis bug #721)

v0.62 2015-01-19  moocow
	* basename fixes dtatw-sanitize-header.perl (was dumping empty basename for -b ./BASENAME calls

v0.62 2015-01-09  moocow
	* added Algorithm::BinarySearch::Vec dependency (for dtatw-get-ddc-attrs.perl)

v0.62 2015-01-06  moocow
	* no --backlink in POD2HTMLFLAGS (ubuntu/debian snafu)
	* fix for goofy text-length explosion on kira (ubuntu server 14.04.1 LTS)
	  - assuming problem was related to printf format sizes and datatype underflow
	  - fix uses PRIu32 macro from inttypes.h to print uint32_t safely
	  - alternate solution uses %u and (uint)ARG , assuming (uint) is at least 32 bits wide

v0.61 2014-12-19  moocow
	* header title extraction fixes

v0.61 2014-12-17  moocow
	* tweaks for ubuntu-server 14.04.1 / perl 5.18.2
	* ignore errors from pod2* utilities

v0.61 2014-12-15  moocow
	* space-normalization for textClass

v0.61 2014-12-12  moocow
	* tcfencode/decode : text/tei+xml adjustments
	* tcfencode.pm : added textSourceType argument for tcfencode object

v0.61 2014-11-28  moocow
	* added tcftokenize doc
	* added tcf2tok target: direct tcf tokenization

src/dtatwUtf8.c
/*
  Basic UTF-8 manipulation routines
  by Jeff Bezanson
  placed in the public domain Fall 2005

  This code is designed to provide the utilities you need to manipulate
  UTF-8 as an internal string encoding. These functions do not perform the
  error checking normally needed when handling UTF-8 data, so if you happen
  to be from the Unicode Consortium you will want to flay me alive.
  I do this because error checking can be performed at the boundaries (I/O),
  with these routines reserved for higher performance on data known to be
  valid.
*/
#include <stdint.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdarg.h>
#ifdef WIN32
#include <malloc.h>
#else
#include <alloca.h>
#endif

#include "dtatwUtf8.h"

static const uint32_t offsetsFromUTF8[6] = {
    0x00000000UL, 0x00003080UL, 0x000E2080UL,
    0x03C82080UL, 0xFA082080UL, 0x82082080UL
};

static const char trailingBytesForUTF8[256] = {
    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1, 1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
    2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2, 3,3,3,3,3,3,3,3,4,4,4,4,5,5,5,5
};

/* returns length of next utf-8 sequence */
int u8_seqlen(const char *s)
{
    return trailingBytesForUTF8[(unsigned int)(unsigned char)s[0]] + 1;
}

/* returns true iff ch is a combining diacritical mark. */
int u8_is_combining(uint32_t ch)
{
  //-- combining diacritical marks; from http://en.wikipedia.org/wiki/Combining_character (Mon, 08 Oct 2012 12:04:13 +0200)
  if ((  ch >= 0x0300 && ch <= 0x036f) // Combining Diacritical Marks (0300–036F), since version 1.0, with modifications in subsequent versions down to 4.1
      ||(ch >= 0x1dc0 && ch <= 0x1dff) // Combining Diacritical Marks Supplement (1DC0–1DFF), versions 4.1 to 5.2
      ||(ch >= 0x20d0 && ch <= 0x20ff) // Combining Diacritical Marks for Symbols (20D0–20FF), since version 1.0, with modifications in subsequent versions down to 5.1
      ||(ch >= 0xfe20 && ch <= 0xfe2f) // Combining Half Marks (FE20–FE2F), versions 1.0, updates in 5.2
      )
    return 1;
  return 0;
}

/* conversions without error checking
   only works for valid UTF-8, i.e. no 5- or 6-byte sequences
   srcsz = source size in bytes, or -1 if 0-terminated
   sz = dest size in # of wide characters

   returns # characters converted
   dest will always be L'\0'-terminated, even if there isn't enough room
   for all the characters.
   if sz = srcsz+1 (i.e. 4*srcsz+4 bytes), there will always be enough space.
*/
int u8_toucs(uint32_t *dest, int sz, const char *src, int srcsz)
{
    uint32_t ch;
    const char *src_end = src + srcsz;
    int nb;
    int i=0;

    while (i < sz-1) {
        nb = trailingBytesForUTF8[(unsigned char)*src];
        if (srcsz == -1) {
            if (*src == 0)
                goto done_toucs;
        }
        else {
            if (src + nb >= src_end)
                goto done_toucs;
        }
        ch = 0;
        switch (nb) {
            /* these fall through deliberately */
        case 3: ch += (unsigned char)*src++; ch <<= 6;
        case 2: ch += (unsigned char)*src++; ch <<= 6;
        case 1: ch += (unsigned char)*src++; ch <<= 6;
        case 0: ch += (unsigned char)*src++;
        }
        ch -= offsetsFromUTF8[nb];
        dest[i++] = ch;
    }
 done_toucs:
    dest[i] = 0;
    return i;
}

/* srcsz = number of source characters, or -1 if 0-terminated
   sz = size of dest buffer in bytes

   returns # characters converted
   dest will only be '\0'-terminated if there is enough space. this is
   for consistency; imagine there are 2 bytes of space left, but the next
   character requires 3 bytes. in this case we could NUL-terminate, but in
   general we can't when there's insufficient space. therefore this function
   only NUL-terminates if all the characters fit, and there's space for
   the NUL as well.
   the destination string will never be bigger than the source string.
*/
int u8_toutf8(char *dest, int sz, const uint32_t *src, int srcsz)
{
    uint32_t ch;
    int i = 0;
    char *dest_end = dest + sz;

    while (srcsz<0 ? src[i]!=0 : i < srcsz) {
        ch = src[i];
        if (ch < 0x80) {
            if (dest >= dest_end)
                return i;
            *dest++ = (char)ch;
        }
        else if (ch < 0x800) {
            if (dest >= dest_end-1)
                return i;
            *dest++ = (ch>>6) | 0xC0;
            *dest++ = (ch & 0x3F) | 0x80;
        }
        else if (ch < 0x10000) {
            if (dest >= dest_end-2)
                return i;
            *dest++ = (ch>>12) | 0xE0;
            *dest++ = ((ch>>6) & 0x3F) | 0x80;
            *dest++ = (ch & 0x3F) | 0x80;
        }
        else if (ch < 0x110000) {
            if (dest >= dest_end-3)
                return i;
            *dest++ = (ch>>18) | 0xF0;
            *dest++ = ((ch>>12) & 0x3F) | 0x80;
            *dest++ = ((ch>>6) & 0x3F) | 0x80;
            *dest++ = (ch & 0x3F) | 0x80;
        }
        i++;
    }
    if (dest < dest_end)
        *dest = '\0';
    return i;
}

/*-- moo: get number of bytes required for representing a wide character \a ch in UTF-8 --*/
int u8_wc_len(uint32_t ch)
{
    if (ch < 0x80) return 1;
    if (ch < 0x800) return 2;
    if (ch < 0x10000) return 3;
    if (ch < 0x110000) return 4;
    return 0;  //-- bad char (?)
}

/*-- moo: get number of bytes required for representing a wide character string \a ws ch in UTF-8 --*/
int u8_ws_len(const uint32_t *src, int srcsz)
{
  int i,len;
  for (i=0,len=0; i < srcsz; i++) {
    len += u8_wc_len(src[i]);
  }
  return len;
}

int u8_wc_toutf8(char *dest, uint32_t ch)
{
    if (ch < 0x80) {
        dest[0] = (char)ch;
        return 1;
    }
    if (ch < 0x800) {
        dest[0] = (ch>>6) | 0xC0;
        dest[1] = (ch & 0x3F) | 0x80;
        return 2;
    }
    if (ch < 0x10000) {
        dest[0] = (ch>>12) | 0xE0;
        dest[1] = ((ch>>6) & 0x3F) | 0x80;
        dest[2] = (ch & 0x3F) | 0x80;
        return 3;
    }
    if (ch < 0x110000) {
        dest[0] = (ch>>18) | 0xF0;
        dest[1] = ((ch>>12) & 0x3F) | 0x80;
        dest[2] = ((ch>>6) & 0x3F) | 0x80;
        dest[3] = (ch & 0x3F) | 0x80;
        return 4;
    }
    return 0;
}

/* charnum => byte offset */
int u8_offset(const char *str, int charnum)
{
    int offs=0;

    while (charnum > 0 && str[offs]) {
        (void)(isutf(str[++offs]) || isutf(str[++offs]) ||
               isutf(str[++offs]) || ++offs);
        charnum--;
    }
    return offs;
}

/* byte offset => charnum */
int u8_charnum(const char *s, int offset)
{
    int charnum = 0, offs=0;

    while (offs < offset && s[offs]) {
        (void)(isutf(s[++offs]) || isutf(s[++offs]) ||
               isutf(s[++offs]) || ++offs);
        charnum++;
    }
    return charnum;
}

/* reads the next utf-8 sequence out of a string, updating an index */
uint32_t u8_nextchar(const char *s, int *i)
{
    uint32_t ch = 0;
    int sz = 0;

    do {
        ch <<= 6;
        ch += (unsigned char)s[(*i)++];
        sz++;
    } while (s[*i] && !isutf(s[*i]));
    ch -= offsetsFromUTF8[sz-1];

    return ch;
}

/*-- moo: reads the next utf-8 sequence out of a string, updating an index, and not exceeding a given length */
uint32_t u8_nextcharn(const char *s, int slen, int *i)
{
    uint32_t ch = 0;
    int sz = 0;

    do {
        ch <<= 6;
        ch += (unsigned char)s[(*i)++];
        sz++;
    } while (*i < slen && !isutf(s[*i]));
    ch -= offsetsFromUTF8[sz-1];

    return ch;
}

/* (moo): return next character; expects NUL-terminated \a s */
uint32_t u8_peek(const char *s)
{
  int i = 0;
  return u8_nextchar(s,&i);
}

/* (moo): return next character; expects NUL-terminated \a s */
uint32_t u8_peekn(const char *s, int slen)
{
  int i = 0;
  return u8_nextcharn(s,slen,&i);
}



/* number of characters */
int u8_strlen(const char *s)
{
    int count = 0;
    int i = 0;

    while (u8_nextchar(s, &i) != 0)
        count++;

    return count;
}

void u8_inc(const char *s, int *i)
{
    (void)(isutf(s[++(*i)]) || isutf(s[++(*i)]) ||
           isutf(s[++(*i)]) || ++(*i));
}

void u8_dec(const char *s, int *i)
{
    (void)(isutf(s[--(*i)]) || isutf(s[--(*i)]) ||
           isutf(s[--(*i)]) || --(*i));
}

int octal_digit(char c)
{
    return (c >= '0' && c <= '7');
}

int hex_digit(char c)
{
    return ((c >= '0' && c <= '9') ||
            (c >= 'A' && c <= 'F') ||
            (c >= 'a' && c <= 'f'));
}

/* assumes that src points to the character after a backslash
   returns number of input characters processed */
int u8_read_escape_sequence(const char *str, uint32_t *dest)
{
    uint32_t ch;
    char digs[9]="\0\0\0\0\0\0\0\0";
    int dno=0, i=1;

    ch = (uint32_t)str[0];    /* take literal character */
    if (str[0] == 'n')
        ch = L'\n';
    else if (str[0] == 't')
        ch = L'\t';
    else if (str[0] == 'r')
        ch = L'\r';
    else if (str[0] == 'b')
        ch = L'\b';
    else if (str[0] == 'f')
        ch = L'\f';
    else if (str[0] == 'v')
        ch = L'\v';
    else if (str[0] == 'a')
        ch = L'\a';
    else if (octal_digit(str[0])) {
        i = 0;
        do {
            digs[dno++] = str[i++];
        } while (octal_digit(str[i]) && dno < 3);
        ch = strtol(digs, NULL, 8);
    }
    else if (str[0] == 'x') {
        while (hex_digit(str[i]) && dno < 2) {
            digs[dno++] = str[i++];
        }
        if (dno > 0)
            ch = strtol(digs, NULL, 16);
    }
    else if (str[0] == 'u') {
        while (hex_digit(str[i]) && dno < 4) {
            digs[dno++] = str[i++];
        }
        if (dno > 0)
            ch = strtol(digs, NULL, 16);
    }
    else if (str[0] == 'U') {
        while (hex_digit(str[i]) && dno < 8) {
            digs[dno++] = str[i++];
        }
        if (dno > 0)
            ch = strtol(digs, NULL, 16);
    }
    *dest = ch;

    return i;
}

/* convert a string with literal \uxxxx or \Uxxxxxxxx characters to UTF-8
   example: u8_unescape(mybuf, 256, "hello\\u220e")
   note the double backslash is needed if called on a C string literal */
int u8_unescape(char *buf, int sz, const char *src)
{
    int c=0, amt;
    uint32_t ch;
    char temp[4];

    while (*src && c < sz) {
        if (*src == '\\') {
            src++;
            amt = u8_read_escape_sequence(src, &ch);
        }
        else {
            ch = (uint32_t)*src;
            amt = 1;
        }
        src += amt;
        amt = u8_wc_toutf8(temp, ch);
        if (amt > sz-c)
            break;
        memcpy(&buf[c], temp, amt);
        c += amt;
    }
    if (c < sz)
        buf[c] = '\0';
    return c;
}

int u8_escape_wchar(char *buf, int sz, uint32_t ch)
{
    if (ch == L'\n')
        return snprintf(buf, sz, "\\n");
    else if (ch == L'\t')
        return snprintf(buf, sz, "\\t");
    else if (ch == L'\r')
        return snprintf(buf, sz, "\\r");
    else if (ch == L'\b')
        return snprintf(buf, sz, "\\b");
    else if (ch == L'\f')
        return snprintf(buf, sz, "\\f");
    else if (ch == L'\v')
        return snprintf(buf, sz, "\\v");
    else if (ch == L'\a')
        return snprintf(buf, sz, "\\a");
    else if (ch == L'\\')
        return snprintf(buf, sz, "\\\\");
    else if (ch < 32 || ch == 0x7f)
        return snprintf(buf, sz, "\\x%hhX", (unsigned char)ch);
    else if (ch > 0xFFFF)
        return snprintf(buf, sz, "\\U%.8X", (uint32_t)ch);
    else if (ch >= 0x80 && ch <= 0xFFFF)
        return snprintf(buf, sz, "\\u%.4hX", (unsigned short)ch);

    return snprintf(buf, sz, "%c", (char)ch);
}

int u8_escape(char *buf, int sz, const char *src, int escape_quotes)
{
    int c=0, i=0, amt;

    while (src[i] && c < sz) {
        if (escape_quotes && src[i] == '"') {
            amt = snprintf(buf, sz - c, "\\\"");
            i++;
        }
        else {
            amt = u8_escape_wchar(buf, sz - c, u8_nextchar(src, &i));
        }
        c += amt;
        buf += amt;
    }
    if (c < sz)
        *buf = '\0';
    return c;
}

char *u8_strchr(char *s, uint32_t ch, int *charn)
{
    int i = 0, lasti=0;
    uint32_t c;

    *charn = 0;
    while (s[i]) {
        c = u8_nextchar(s, &i);
        if (c == ch) {
            return &s[lasti];
        }
        lasti = i;
        (*charn)++;
    }
    return NULL;
}

char *u8_memchr(char *s, uint32_t ch, size_t sz, int *charn)
{
    int i = 0, lasti=0;
    uint32_t c;
    int csz;

    *charn = 0;
    while ((size_t)i < sz) {
        c = csz = 0;
        do {
            c <<= 6;
            c += (unsigned char)s[i++];
            csz++;
        } while ((size_t)i < sz && !isutf(s[i]));
        c -= offsetsFromUTF8[csz-1];

        if (c == ch) {
            return &s[lasti];
        }
        lasti = i;
        (*charn)++;
    }
    return NULL;
}

int u8_is_locale_utf8(const char *locale)
{
    /* this code based on libutf8 */
    const char* cp = locale;

    for (; *cp != '\0' && *cp != '@' && *cp != '+' && *cp != ','; cp++) {
        if (*cp == '.') {
            const char* encoding = ++cp;
            for (; *cp != '\0' && *cp != '@' && *cp != '+' && *cp != ','; cp++)
                ;
            if ((cp-encoding == 5 && !strncmp(encoding, "UTF-8", 5))
                || (cp-encoding == 4 && !strncmp(encoding, "utf8", 4)))
                return 1; /* it's UTF-8 */
            break;
        }
    }
    return 0;
}

int u8_vprintf(const char *fmt, va_list ap)
{
    int cnt, sz=0;
    char *buf;
    uint32_t *wcs;

    sz = 512;
    buf = (char*)alloca(sz);
 try_print:
    cnt = vsnprintf(buf, sz, fmt, ap);
    if (cnt >= sz) {
        buf = (char*)alloca(cnt - sz + 1);
        sz = cnt + 1;
        goto try_print;
    }
    wcs = (uint32_t*)alloca((cnt+1) * sizeof(uint32_t));
    cnt = u8_toucs(wcs, cnt+1, buf, cnt);
    printf("%ls", (wchar_t*)wcs);
    return cnt;
}

int u8_printf(const char *fmt, ...)
{
    int cnt;
    va_list args;

    va_start(args, fmt);

    cnt = u8_vprintf(fmt, args);

    va_end(args);
    return cnt;
}
